### DISCUSSION TITLE
Formatting entries to indicate which of them have associated notes and/or PDFs

### DISCUSSION THREAD

**Discussion by apc** [emacs-citar/citar#93] [Q&A] [answered]
I've tried setting the value of `bibtex-actions-template-suffix` to include `${=has-pdf=:1}${=has-note=:1}`:. 

```
bibtex-actions-template-suffix is a variable defined in ‘bibtex-actions.el’.
Its value is
((t . "          ${=key=:15} ${=has-pdf=:1}${=has-note=:1}   ${=type=:12}    ${tags:*}"))
Original value was 
((t . "          ${=key=:15}    ${=type=:12}    ${tags:*}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry suffix.
When combined wiht the main template, the same string is used for
display and for search.
```

I've checked the value of `bibtex-actions-template` and it is what I assume it should be: 

```
bibtex-actions-template is a variable defined in ‘bibtex-actions.el’.
Its value is ((t . "${author:20}   ${title:48}   ${year:4}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry.
When combined with the suffix, the same string is used for
display and for search.
```

I now see the symbols indicating presence of notes and/or pdfs. But somehow they appear twice and break the alignment of the entries: 

<img width="1606" alt="image" src="https://user-images.githubusercontent.com/350504/114894458-a44fef00-9ddc-11eb-830c-49d7cbc28bea.png">

Any advice on how to adjust this? 

(Thanks again!)

I'm on Emacs 27.2.

**Comment by bdarcus** [comment, marked as answer]
Edit: see [below](https://github.com/bdarcus/bibtex-actions/discussions/93#discussioncomment-616792); it's a bug in your Emacs 27.2.

--------

What happens if you change the suffix to this:

``` elisp
((t . "          ${=key=:15} ${=has-note=:1}   ${=type=:12}    ${tags:*}"))
```

... and then the following?

``` elisp
((t . "          ${=key=:15} ${=has-pdf=:1}${=has-note=:1}   ${=type=:12}"))
```

PS - depending on your answer, I may create an issue of this. It's clearly a bug somewhere; just not sure if it's here or not.

**Reply by apc** [reply]
Same thing. 

<img width="1455" alt="image" src="https://user-images.githubusercontent.com/350504/114905563-16c5cc80-9de7-11eb-904f-e801127e0569.png">

```
bibtex-actions-template-suffix is a variable defined in ‘bibtex-actions.el’.
Its value is
((t . "          ${=key=:15} ${=has-pdf=:1}${=has-note=:1}   ${=type=:12}"))
Original value was 
((t . "          ${=key=:15}    ${=type=:12}    ${tags:*}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry suffix.
When combined wiht the main template, the same string is used for
display and for search.
```

**Reply by bdarcus** [reply]
The same thing in both cases? Eg even without the pdf showing in your suffix, it shows on the very left?

**Reply by bdarcus** [reply]
Here's what mine looks like on 28, with the same template setting.

![Screenshot from 2021-04-15 12-49-09](https://user-images.githubusercontent.com/1134/114907422-0c0c3700-9de9-11eb-8c03-2f2972396086.png)

**Comment by apc** [comment]
Oh. Sorry, I misread your first question. I’ll try again. Away from my computer now.

**Reply by apc** [reply]
OK, this is what I get with the other template: 

<img width="1465" alt="image" src="https://user-images.githubusercontent.com/350504/114908853-8093a580-9dea-11eb-8c04-a1414d115279.png">

```
bibtex-actions-template-suffix is a variable defined in ‘bibtex-actions.el’.
Its value is
((t . "          ${=key=:15} ${=has-note=:1}   ${=type=:12}    ${tags:*}"))
Original value was 
((t . "          ${=key=:15}    ${=type=:12}    ${tags:*}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry suffix.
When combined wiht the main template, the same string is used for
display and for search.
```

**Reply by bdarcus** [reply]
Let me tell you what I'm thinking:

The first symbol is rendered using the affixation-function, which should not work at all on Emacs 27.

So my hypothesis is it works (maybe they backported it more recently?), but not properly.

The easiest way to confirm is for you to load the bibtex-actions.el file, comment out this line, and then eval buffer.

``` elisp
 (affixation-function . bibtex-actions--affixation)
```

If I'm right, that symbol will go away, and the alignment will work correctly.

If not, I don't ATM have any idea how it could render like that, so will have to think about it.

**Reply by apc** [reply]
Bingo! That got rid of it. 

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/350504/114910033-660dfc00-9deb-11eb-8631-e26c9d6ffe0c.png">

**Reply by bdarcus** [reply]
So it's a bug in the version of Emacs that you're running, I guess.

I'm going to try to ping @minad to see if he knows anything about affixation being backported to Emacs 27, but not working correctly. Not sure if that will work in github discussions though.

My understanding is previous to 28 versions will ignore the affixation-function, but that's clearly not the case here.

**Reply by apc** [reply]
Any thoughts on what to do for a temporary fix? Leave that line commented out? (I imagine I need to remove the `elc` file for this to take effect upon restart?)

**Reply by apc** [reply]
OK, I just did that (`byte-force-recompile`) after commenting out that line, and it works.

**Reply by bdarcus** [reply]
That's good, but obviously not ideal for you long-term.

I just checked the 27 branch source code, and there's no mention of affixation in the minibuffer.el file.

So now I'm really confused.

It is possible your version is somehow, um, idiosyncratic?

**Reply by apc** [reply]
My version of Emacs?? I got it from here: https://github.com/railwaycat/homebrew-emacsmacport

I used the `--with-modules` and `--with-natural-title-bar` options. Aside from that, not sure what other idiosyncrasies could have snuck in!

**Reply by bdarcus** [reply]
Could you try replacing the commented line with this?

``` elisp
                       (if (eq action 'metadata 'affixation-function)
                        (affixation-function . bibtex-actions--affixation))
```

If that works, I might be able to merge it, but would want to take a closer look to see if anything explains the behavior.

**Reply by apc** [reply]
Yes, it works!

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/350504/114917476-fcdeb680-9df3-11eb-9226-0d1545d31da4.png">

**Reply by bdarcus** [reply]
Alright, I pinged an expert on this to get his advice on the issue I created. But at least you now have what you were expecting!
